docs/manuscripts/cosmicqc_paper_v1/figure_3/nbconverted/compute_silho_scores.py

```diff
@@ -0,0 +1,141 @@
+)
+post_QC_umap_df = pd.read_parquet(
+    "../figure_3/umap_embeddings/post_QC_umap_embeddings.parquet"
+)
+
+# Set min cluster size for HDBSCAN
+min_cluster_size = 100
+
+
+# ## Compute individual Silhouette scores
+
+# ### Pre-QC UMAP score
+
+# In[3]:
+
+
+# Calculate and print silhouette score for pre-QC datasets
+pre_X = pre_QC_umap_df[["UMAP0", "UMAP1"]].values
+
+# Run HDBSCAN
+clusterer = hdbscan.HDBSCAN(min_cluster_size=min_cluster_size)
+pre_cluster_labels = clusterer.fit_predict(pre_X)
+
+# Info about clusters
+print("Unique cluster labels (-1 is noise):", np.unique(pre_cluster_labels))
+print("Cluster persistence:", clusterer.cluster_persistence_)
+
+# Silhouette score (exclude noise points labeled as -1)
+mask = pre_cluster_labels != -1
+X_clustered = pre_X[mask]
+labels_clustered = pre_cluster_labels[mask]
+pre_QC_score = silhouette_score(X_clustered, labels_clustered)
+print("Silhouette score (excluding noise):", pre_QC_score)
+
+
+# ### Post-QC UMAP score
+
+# In[4]:
+
+
+# Calculate and print silhouette scores for post-QC datasets
+post_X = post_QC_umap_df[["UMAP0", "UMAP1"]].values
+
+# Run HDBSCAN
+clusterer = hdbscan.HDBSCAN(min_cluster_size=min_cluster_size)
+post_cluster_labels = clusterer.fit_predict(post_X)
+
+# Info about clusters
+print("Unique cluster labels (-1 is noise):", np.unique(post_cluster_labels))
+print("Cluster persistence:", clusterer.cluster_persistence_)
+
+# Silhouette score (exclude noise points labeled as -1)
+mask = post_cluster_labels != -1
+post_X_clustered = post_X[mask]
+labels_clustered = post_cluster_labels[mask]
+post_QC_score = silhouette_score(post_X_clustered, labels_clustered)
+print("Silhouette score (excluding noise):", post_QC_score)
+
+
+# # Visualize where passing or failing cells are within each cluster
+
+# In[5]:
+
+
+# Attach cluster labels to full dataframe
+pre_QC_clusters_df = pre_QC_umap_df.copy()
+pre_QC_clusters_df["cluster"] = pre_cluster_labels
+
+# Drop noise
+pre_QC_clusters_df = pre_QC_clusters_df[pre_QC_clusters_df["cluster"] != -1]
+
+# Convert labels to string for nicer plotting
+pre_QC_clusters_df["cluster"] = pre_QC_clusters_df["cluster"].astype(str)
+
+# Print number of cells in each cluster
+print(pre_QC_clusters_df["cluster"].value_counts().sort_index())
+
+# Calculate proportions for each cluster
+cluster_counts = (
+    pre_QC_clusters_df.groupby(["cluster", "Metadata_QC_status"])
+    .size()
+    .unstack(fill_value=0)
+)
+cluster_props = cluster_counts.div(cluster_counts.sum(axis=1), axis=0)
+
+# Plot
+ax = cluster_props.plot(
+    kind="bar", stacked=True, color=["tomato", "mediumseagreen"], figsize=(8, 5)
+)
+ax.set_ylabel("Proportion of Cells")
+ax.set_xlabel("Cluster Label")
+ax.set_title("Proportion of Cells Passed/Failed QC by Cluster")
+ax.legend(["Failed QC", "Passed QC"], title="QC Status")
+plt.tight_layout()
+plt.savefig(output_dir / "pre_QC_cluster_qc_status_proportions.png", dpi=300)
+plt.show()
```

Comment:
This panel looks good to me, but I think my comments on the manuscript will demonstrate the Silhouette score analysis is not appropriately testing our hypothesis of the pre vs. post QC clustering improvements.

Reply:
Updated!